### PR TITLE
MySQL backend, interval implementations, inference improvements

### LINKED ISSUE #1164: Support external tables for clickhouse
Clickhouse can receive external tables when executing a query, see [external data](https://clickhouse.yandex/docs/en/table_engines/external_data.html) in the documentation.
The external tables can passed to `clickhouse-driver` in the [following way](https://github.com/mymarilyn/clickhouse-driver/blob/94e083589a83ffc289ef2468ab110c141d8cc241/tests/test_external_tables.py#L6).

Ibis currently doesn't support passing extra parameters down to the query execution, however there is a currently unused [params](https://github.com/ibis-project/ibis/blob/master/ibis/client.py#L208) argument.

What would be the best way to push down external tables (received as pandas dataframes) down to the execution? 
- just wire the params argument through the method calls
- set as context property for `Query` object?

### LINKED ISSUE #1224: MySQL backend
I guess mysql would be a nice to have backend.

Should it be based on the alchemy backend or the [sql compiler](https://github.com/ibis-project/ibis/blob/master/ibis/sql/compiler.py) (like impala, clickhouse and bigquery)?

### LINKED ISSUE #1249: Deprecate timedelta API in favor of intervals

### LINKED ISSUE #1251: interval operations in postgres

### LINKED ISSUE #1260: Date truncate
Truncate is only implemented for timestamp whereas date should support that too.

### LINKED ISSUE #1262: isinf/isnan backend implementations
In no particular order:

- [ ] postgres
- [ ] sqlite
- [ ] clickhouse
- [ ] impala
- [ ] bigquery
- [ ] pandas

### LINKED ISSUE #1323: Clickhouse table insert
Probably the schema conversion and argument validation can be generalized to base `SQLClient`

### PR DESCRIPTION
**Pull request by kszucs**
- [x] MySQL backend, including a couple of window operations if the server supports (resolves #1224)
- [x] Interval implementations (resolves #1251)
- [x] Temporal truncate implementations (resolves #1260)
- [x] Deprecate timedelta API (resolves #1249)
- [x] Clickhouse external tables (resolves #1164)
- [x] Clickhouse table insert (resolves #1323)
- [x] IsInf / IsNan implementations (partially resolves #1262)
- [x] Client version property
- [x] `dtype.to_pandas()`, `schema.to_pandas()`
- [x] Csv reading improvements
- [x] Numpy and Pandas datatype handling moved under pandas (except for [this](https://github.com/ibis-project/ibis/pull/1292/files#diff-0b9f3a78e49f59ac30cc4ca44949b25fR1169) one)
- Started to move tests cases under `all` (numeric, param) 
- Condensed backend test suite (I hope @cpcloud doesn't mind)
- Minor test fixes, caused mostly by incorrect dtypes
- Reduced duplicated code in alchemy backends

|  | Csv | Parquet | Pandas | Sqlite | MySQL | Postgres | Clickhouse | Impala | BigQuery |
| --: | :-: | :-: | :-: | :-: | :-: | :-: | :-: | :-: | :-: |
| Interval | X | X | X  |   |   | X | X | X |  |
| IntegerToInterval | X | X | X  |   |   | X | X | X |  |
| DateTruncate | X | X | X | X | X | X | X | X |  |
| IsInf / IsNan | X | X | X |   |   | X | X | X |  |

### CODE REVIEW

**Comment by kszucs** [comment]
@cpcloud Please review! The build are going to be green (probably), but I expect a couple of change requests :)

**Comment by kszucs** [comment]
@cpcloud Rebase done

File: ibis/sql/compiler.py [hunk 1]
```diff
@@ -1023,7 +1023,7 @@ def make_alias(self, expr):
         alias = 't%d' % i
         self.set_ref(expr, alias)
 
-    def need_aliases(self):
+    def need_aliases(self, expr=None):
```
Comment (cpcloud) [inline ibis/sql/compiler.py]: Why this change?

File: ibis/sql/mysql/compiler.py [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+    ops.StringFind: _string_find,
+    ops.StringLength: unary(sa.func.length),
+    # ops.GroupConcat: fixed_arity('concat_ws', 2),
+    ops.Lowercase: unary(sa.func.lower),
+    ops.Uppercase: unary(sa.func.upper),
+    ops.Strip: unary(sa.func.trim),
+    ops.LStrip: unary(sa.func.ltrim),
+    ops.RStrip: unary(sa.func.rtrim),
+    ops.LPad: fixed_arity(sa.func.lpad, 3),
+    ops.RPad: fixed_arity(sa.func.rpad, 3),
```
Comment (cpcloud) [inline ibis/sql/mysql/compiler.py]: We should factor these out

File: ibis/sql/alchemy.py [hunk 3]
```diff
@@ -432,6 +439,106 @@ def _string_like(t, expr):
...
+    new_op = klass(*op.args)
+    new_expr = expr._factory(new_op, name=expr._name)
+
+    if type(new_op) in translator._rewrites:
+        new_expr = translator._rewrites[type(new_op)](new_expr)
+
+    return L.windowize_function(new_expr, win)
+
+
+def _window(t, expr):
```
Comment (cpcloud) [inline ibis/sql/alchemy.py]: We desperately need to refactor this.

File: ibis/sql/postgres/tests/test_functions.py [hunk 4]
```diff
@@ -510,14 +510,14 @@ def test_numeric_builtins_work(alltypes, df):
             lambda t: (t.double_col > 20).ifelse(10, -20),
             lambda df: pd.Series(
                 np.where(df.double_col > 20, 10, -20),
-                dtype='int64'
+                dtype='int8'
```
Comment (cpcloud) [inline ibis/sql/postgres/tests/test_functions.py]: How does this pass?

File: ibis/tests/all/conftest.py [hunk 5]
```diff
@@ -68,3 +67,40 @@ def analytic_alltypes(alltypes):
...
+# @pytest.fixture(scope='session')
+# def rewrites(translator):
+#     return translator._rewrites
+
+
+# @pytest.fixture(scope='session')
+# def valid_operations(registry, rewrites, backend):
+#     return (
+#         frozenset(registry) | frozenset(rewrites)
+#     ) - backend.additional_skipped_operations
```
Comment (cpcloud) [inline ibis/tests/all/conftest.py]: Are these unused? If so, then please remove them.

File: ibis/tests/all/conftest.py [hunk 6]
```diff
@@ -1,63 +1,62 @@
...
+@pytest.fixture(params=[
+    pytest.param(Csv, marks=pytest.mark.csv),
+    pytest.param(Parquet, marks=pytest.mark.parquet),
+    pytest.param(Pandas, marks=pytest.mark.pandas),
+    pytest.param(SQLite, marks=pytest.mark.sqlite),
+    pytest.param(Postgres, marks=pytest.mark.postgres),
+    pytest.param(MySQL, marks=pytest.mark.mysql),
+    pytest.param(Clickhouse, marks=pytest.mark.clickhouse),
+    pytest.param(BigQuery, marks=pytest.mark.bigquery),
+    pytest.param(Impala, marks=pytest.mark.impala)
```
Comment (cpcloud) [inline ibis/tests/all/conftest.py]: Ok, that is awesome. I was wondering if there was a sane way to mark parameterized fixtures.

**Comment by cpcloud** [comment]
This is awesome. Thanks @kszucs, I will try to get this merged as you can address the comments. Apologies for the delay :)

File: ci/datamgr.py [hunk 7]
```diff
@@ -1,34 +1,39 @@
...
 
 from toolz import dissoc
 from plumbum import local
 from plumbum.cmd import curl, psql
 
+try:
+    from pathlib import Path
+except ImportError:
+    # py2 compat
+    from pathlib2 import Path
```
[the remark this answers is not in the document]
Comment (kszucs) [reply ci/datamgr.py]: Actually datamgr doesn't import ibis. Shouldn't we keep that way?

File: ci/requirements-dev-3.5.yml [hunk 8]
```diff
@@ -6,20 +6,25 @@ dependencies:
...
   - python=3.5
   - python-graphviz
+  - pyarrow>=0.6.0
   - regex
   - six
   - sqlalchemy>=1.0.0,<1.1.15
   - thrift<=0.9.3
   - thriftpy<=0.3.9
   - toolz
+  - clickhouse-cityhash
```
[the remark this answers is not in the document]
Comment (kszucs) [reply ci/requirements-dev-3.5.yml]: This is for client side compression, not mandatory though.

File: ci/requirements-dev-3.6.yml [hunk 9]
```diff
@@ -6,11 +6,14 @@ dependencies:
   - cmake
   - flake8
   - graphviz
+  - requests
   - impyla>=0.13.7
+  - lz4
```
[the remark this answers is not in the document]
Comment (kszucs) [reply ci/requirements-dev-3.6.yml]: Same, clickhouse-driver compression.

File: ibis/__init__.py [hunk 10]
```diff
@@ -14,67 +14,61 @@
...
+from ibis.filesystems import HDFS, WebHDFS
+
 # __all__ is defined
 from ibis.expr.api import *
 
-try:
+
+halt_ordering()
+import ibis.pandas.api as pandas
+restart_ordering()
```
[the remark this answers is not in the document]
Comment (kszucs) [reply ibis/__init__.py]: You tell me :) It seems to work.

File: ibis/clickhouse/client.py [hunk 11]
```diff
@@ -2,69 +2,137 @@
...
+    'String': dt.String,
+    'FixedString': dt.String,
+    'Date': dt.Date,
+    'DateTime': dt.Timestamp
+}
+_ibis_dtypes = {v: k for k, v in _clickhouse_dtypes.items()}
+_ibis_dtypes[dt.String] = 'String'
+
+
+class ClickhouseDataType(object):
```
[the remark this answers is not in the document]
Comment (kszucs) [reply ibis/clickhouse/client.py]: Clickhouse doesn't depend on alchemy.

[the remark this answers is not in the document]
Comment (kszucs) [reply ibis/clickhouse/client.py]: I've created an issue for that https://github.com/mymarilyn/clickhouse-driver/issues/22, no reaction yet. I could refactor this piece of code once `clickhouse-driver` ships column specification objects instead of strings.

File: ibis/clickhouse/client.py [hunk 12]
```diff
@@ -96,12 +164,23 @@ def close(self):
         """Close Clickhouse connection and drop any temporary objects"""
         self.con.disconnect()
 
-    def _execute(self, query):
+    def _execute(self, query, external_tables=[], results=True):
```
[the remark this answers is not in the document]
Comment (kszucs) [reply ibis/clickhouse/client.py]: Rookie mistake :)

File: ibis/clickhouse/compiler.py [hunk 13]
```diff
@@ -51,6 +51,14 @@ class ClickhouseQueryContext(comp.QueryContext):
     def _to_sql(self, expr, ctx):
         return to_sql(expr, context=ctx)
 
+    def need_aliases(self, expr=None):
+        from ibis.clickhouse.client import ClickhouseExternalTable
+
+        if isinstance(expr.op(), ClickhouseExternalTable):
```
[the remark this answers is not in the document]
Comment (kszucs) [reply ibis/clickhouse/compiler.py]: Because clickhouse throws an error in that case, this is just a workaround.

File: ibis/client.py [hunk 14]
```diff
@@ -46,6 +43,7 @@ def __init__(self, client, ddl):
             ddl, 'parent_expr', getattr(ddl, 'table_set', None)
         )
 
+        self.ddl = ddl
```
[the remark this answers is not in the document]
Comment (kszucs) [reply ibis/client.py]: In ClickhouseQuery when extracting external tables from the expression: https://github.com/ibis-project/ibis/pull/1292/files#diff-cb1dd568c3e11460210845acec54bc13R95

File: ibis/client.py [hunk 15]
```diff
@@ -66,22 +64,14 @@ def _wrap_result(self, result):
...
-        return pd.DataFrame(cols, columns=names)
-
-    def _db_type_to_dtype(self, db_type, column):
         raise NotImplementedError
 
+    def schema(self):
+        try:
+            return self.expr.schema()
+        except AttributeError:
+            return sch.schema([(self.expr.get_name(), self.expr.type())])
```
[the remark this answers is not in the document]
Comment (kszucs) [reply ibis/client.py]: I've created one, but involves calling private api: tests/all/test_client.py

File: ibis/compat.py [hunk 16]
```diff
@@ -89,6 +90,13 @@ def maketrans(x, y):
 integer_types = six.integer_types + (np.integer,)
 
 
+try:
+    from pathlib import Path  # noqa: F401
+except ImportError:
+    # py2 compat
+    from pathlib2 import Path  # noqa: F401
```
[the remark this answers is not in the document]
Comment (kszucs) [reply ibis/compat.py]: Do You have a shortcut for it? :)

File: ibis/expr/api.py [hunk 17]
```diff
@@ -2023,16 +2002,20 @@ def _timestamp_truncate(arg, unit):
...
+      'ms': millisecond
+      'us': microsecond
+      'ns': nanosecond
 
     Returns
     -------
     truncated : timestamp
     """
-    return _ops.Truncate(arg, unit).to_expr()
+    return _ops.TimestampTruncate(arg, unit).to_expr()
```
[the remark this answers is not in the document]
Comment (kszucs) [reply ibis/expr/api.py]: Removed the second one.

File: ibis/expr/tests/test_temporal.py [hunk 18]
```diff
@@ -3,20 +3,48 @@
...
+    api.quarter,
+    api.month,
+    api.week,
+    api.day,
+    api.hour,
+    api.minute,
+    api.second,
+    api.millisecond,
+    api.microsecond,
+    api.nanosecond
```
[the remark this answers is not in the document]
Comment (kszucs) [reply ibis/expr/tests/test_temporal.py]: Yes: https://github.com/ibis-project/ibis/issues/1249

File: ibis/expr/types.py [hunk 19]
```diff
@@ -1425,7 +1435,7 @@ def literal(value, type=None):
     if type is not None:
         try:
             # check that dtype is implicitly castable to explicitly given dtype
-            dtype = dtype.cast(type)
+            dtype = dtype.cast(type, value=value)
```
[the remark this answers is not in the document]
Comment (kszucs) [reply ibis/expr/types.py]: You've added an explicit value kwarg to `DataType.cast` and `DataType.castable` but in another PR: https://github.com/ibis-project/ibis/pull/1303/files#diff-0b9f3a78e49f59ac30cc4ca44949b25fR100

File: ibis/file/csv.py [hunk 20]
```diff
@@ -1,45 +1,24 @@
...
-
-    dt.Date: 'datetime64[D]',
-}
-
-
-def ibis_schema_to_pandas_dtypes(schema):
-    items = zip(schema.names, schema.types)
-    return {name: _IBIS_TO_PANDAS_DTYPE[type(t)] for name, t in items}
+    return pd.read_csv(str(path), dtype=dtypes, parse_dates=dates,
+                       encoding='utf-8', **kwargs)
```
[the remark this answers is not in the document]
Comment (kszucs) [reply ibis/file/csv.py]: Currently each backend on each python versions gives the `string_col` as an unicode column - which requires the utf8 encoding in case of CSV.

OTOH Have You seen my name? :)

File: ibis/impala/tests/test_exprs.py [hunk 21]
```diff
@@ -376,7 +376,7 @@ def test_numeric_unary_builtins(self):
 
     def test_log_other_bases(self):
         cases = [
-            (self.table.double_col.log(5), 'log(`double_col`, 5)')
+            (self.table.double_col.log(5), 'log(5, `double_col`)')
```
[the remark this answers is not in the document]
Comment (kszucs) [reply ibis/impala/tests/test_exprs.py]: If I recall correctly the latter one. Base comes first according to impala's doc: https://www.cloudera.com/documentation/enterprise/5-8-x/topics/impala_math_functions.html

File: ibis/impala/tests/test_exprs.py [hunk 22]
```diff
@@ -1234,25 +1236,23 @@ def test_decimal_builtins_2(self):
...
+            (dc.fillna(0), 5.245),
 
             (dc.exp(), 189.6158),
             (dc.log(), 1.65728),
             (dc.log2(), 2.39094),
             (dc.log10(), 0.71975),
             (dc.sqrt(), 2.29019),
-            (dc.zeroifnull(), Decimal('5.245')),
-            (-dc, Decimal('-5.245'))
+            (dc.zeroifnull(), 5.245),
```
[the remark this answers is not in the document]
Comment (kszucs) [reply ibis/impala/tests/test_exprs.py]: I've changed that behaviour: https://github.com/ibis-project/ibis/pull/1292/files/dc7459bcf5c9d47ac9424d915f9ab9d89e3df60a#diff-84496f511316e2bf7b533a0c979ca954R43

Actually I don't know how to precisely represent decimal types within pandas and impala was the only backend which utilized `Decimal`.

How should I handle ibis decimal to pandas -> `?`

File: ibis/pandas/client.py [hunk 23]
```diff
@@ -109,6 +218,10 @@ def compile(self, expr, *args, **kwargs):
     def database(self, name=None):
         return PandasDatabase(name, self)
 
+    @property
+    def version(self):
+        return tuple(pd.__version__.split('.'))
```
[the remark this answers is not in the document]
Comment (kszucs) [reply ibis/pandas/client.py]: I suggest to use setuptools' `parse_version` which uses `packaging`, see: https://stackoverflow.com/questions/11887762/compare-version-strings-in-python

[the remark this answers is not in the document]
Comment (kszucs) [reply ibis/pandas/client.py]: I've ended up creating a `parse_version` in compat.

File: ibis/pandas/execution/strings.py [hunk 24]
```diff
@@ -23,7 +23,7 @@
 
 @execute_node.register(ops.StringLength, pd.Series)
 def execute_string_length_series(op, data, **kwargs):
-    return data.str.len()
+    return data.str.len().astype('int32')
```
[the remark this answers is not in the document]
Comment (kszucs) [reply ibis/pandas/execution/strings.py]: The output types will be incorrect, therefore otherwise it breaks the tests.
Also `StringLenght` should return `int32` defined in https://github.com/ibis-project/ibis/blob/master/ibis/expr/operations.py#L622

File: ibis/pandas/execution/strings.py [hunk 25]
```diff
@@ -223,7 +223,7 @@ def execute_group_concat_series_gb_mask(
 
 @execute_node.register(ops.StringAscii, pd.Series)
 def execute_string_ascii(op, data, **kwargs):
-    return data.map(ord)
+    return data.map(ord).astype('int32')
```
[the remark this answers is not in the document]
Comment (kszucs) [reply ibis/pandas/execution/strings.py]: Same as above, see `StringAscii` https://github.com/ibis-project/ibis/blob/master/ibis/expr/operations.py#L627

File: ibis/pandas/execution/tests/test_strings.py [hunk 26]
```diff
@@ -10,7 +10,7 @@
 @pytest.mark.parametrize(
     ('case_func', 'expected_func'),
     [
-        (lambda s: s.length(), lambda s: s.str.len()),
+        (lambda s: s.length(), lambda s: s.str.len().astype('int32')),
```
[the remark this answers is not in the document]
Comment (kszucs) [reply ibis/pandas/execution/tests/test_strings.py]: Same answer as above :)

File: ibis/sql/compiler.py (hunk 1 above)
Comment (kszucs) [reply ibis/sql/compiler.py]: Because in case of clickhouse it depends on the expression too not just the dialect.

File: ibis/sql/postgres/tests/test_functions.py (hunk 4 above)
Comment (kszucs) [reply ibis/sql/postgres/tests/test_functions.py]: Because each vaule is between int8's boundaries so the expression's dtype is `dt.int8` instead of int64.

File: ibis/tests/all/conftest.py (hunk 6 above)
Comment (kszucs) [reply ibis/tests/all/conftest.py]: I've almost written a pytest plugin :)

File: ibis/clickhouse/operations.py [hunk 27]
```diff
@@ -578,6 +601,7 @@ def _string_like(translator, expr):
 
     ops.Cast: _cast,
 
+    # for more than 2 args this should be arrayGreatest|Least(array([]))
```
[the remark this answers is not in the document]
Comment (kszucs) [reply ibis/clickhouse/operations.py]: The main issue here is `least` and `greatest` don't support varargs, I've extended the comment.

File: ci/datamgr.py [hunk 28]
```diff
@@ -77,23 +90,47 @@ def cli():
               default='https://storage.googleapis.com/ibis-ci-data')
 @click.option('-d', '--directory', default=SCRIPT_DIR)
 def download(base_url, directory, name):
-    if not os.path.exists(directory):
-        os.mkdir(directory)
+    directory = Path(directory)
+    if not directory.exists():
+        directory.mkdir()
```
[the remark this answers is not in the document]
Comment (kszucs) [reply ci/datamgr.py]: I've tried, but it fails on 3.4 due to exist_ok has been added in 3.5.

File: ibis/__init__.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (kszucs) [reply ibis/__init__.py]: I've extended it, here are the import times (via test_import_time):

Before: 1.78s 
After: 1.63

Roughly 0.1s speedup

**Comment by cpcloud** [comment]
@kszucs Looks like there's a MySQL syntax error on the appveyor builds: https://ci.appveyor.com/project/cpcloud/ibis-xh5g1/build/1.0.1473/job/yck3ybcdk1v0ue4h

File: ibis/tests/util.py [hunk 29]
```diff
@@ -33,12 +31,15 @@ def assert_equal(left, right):
...
     @functools.wraps(f)
     def wrapper(backend, *args, **kwargs):
         try:
             return f(backend, *args, **kwargs)
-        except (com.OperationNotDefinedError, com.UnsupportedBackendType) as e:
-            pytest.skip('{} using {}'.format(e, backend.__name__))
+        except (com.OperationNotDefinedError,
+                com.UnsupportedBackendType,
+                com.TranslationError,
+                NotImplementedError) as e:
```
Comment (cpcloud) [inline ibis/tests/util.py]: Are we sure that none of these skips are spurious?

**Comment by kszucs** [comment]
@cpcloud Because it runs the window tests however that mysql version doesn't support. I'm currently fixing the version checking...

File: ibis/tests/util.py (hunk 29 above)
Comment (kszucs) [reply ibis/tests/util.py]: We can be sure after eliminating all of the TranslationError's and NotImplementedError.
For example multipledispatch raises NotImplementedError on missing signature (in our case missing implementation).

Comment (cpcloud) [reply ibis/tests/util.py]: cool

File: ibis/clickhouse/client.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (cpcloud) [reply ibis/clickhouse/client.py]: Cool, that sounds good to me, can you create an issue on our side as well.

File: ibis/clickhouse/compiler.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (cpcloud) [reply ibis/clickhouse/compiler.py]: Ok, fine by me.

File: ibis/clickhouse/operations.py (hunk 27 above)
[the remark this answers is not in the document]
Comment (cpcloud) [reply ibis/clickhouse/operations.py]: Right, got it. Was just pointing out a thing to keep in mind :)

File: ibis/compat.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (cpcloud) [reply ibis/compat.py]: Ha! :) My mistake.

File: ibis/sql/compiler.py (hunk 1 above)
Comment (cpcloud) [reply ibis/sql/compiler.py]: Ok. We should document this aspect of the design and see if we can generalize it so that people writing a new backend can ignore this detail if they don't need it.

Comment (cpcloud) [reply ibis/sql/compiler.py]: I will do that in my design doc writeup

**Comment by cpcloud** [comment]
Ok, this looks good to me. Bombs away.
